File: app/src/main/java/io/github/getsmp/lorforandroid/ui/section/news/NewsItemFactory.java
/*
 * Copyright 2016 getsmp
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.getsmp.lorforandroid.ui.section.news;

import android.text.Html;

import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.List;

import io.github.getsmp.lorforandroid.ui.section.ItemCommon;
import io.github.getsmp.lorforandroid.ui.section.ItemFactory;
import io.github.getsmp.lorforandroid.util.StringUtils;

public class NewsItemFactory implements ItemFactory {
    @Override
    public void prepareItems(Element body, List items) {
        Elements articles = body.select("article");
        for (Element article : articles) {
            if (article.hasClass("mini-news")) {
                // Mini-news article
                items.add(new MiniNewsItem(
                        article.select("a[href^=/news/]").first().attr("href").substring(1),
                        Html.fromHtml(article.select("a[href^=/news/]").first().ownText()).toString(),
                        Html.fromHtml(article.select("a").first().nextSibling().toString()).toString().replaceAll("[()]", "")
                ));
            } else {
                // Standard article
                items.add(new ItemCommon(
                        article.select("h2 > a[href^=/news/]").first().attr("href").substring(1),
                        Html.fromHtml(article.select("h2 > a[href^=/news/]").first().ownText()).toString(),
                        StringUtils.removeSectionName(article.select("div.group").first().text()),
                        StringUtils.tagsFromElements(article.select("a.tag")),
                        article.select("time").first().ownText(),
                        article.select("a[itemprop^=creator], div.sign:contains(anonymous)").first().ownText().replace(" ()", ""),
                        article.select("div.nav > a[href$=#comments]:eq(0)").first().ownText()
                ));
            }
        }
    }
}

File: app/src/main/java/io/github/getsmp/lorforandroid/ui/section/news/NewsItemFactory.kt
/*
 * Copyright (c) 2016 getsmp
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package io.github.getsmp.lorforandroid.ui.section.news

import android.text.Html
import io.github.getsmp.lorforandroid.ui.section.Item
import io.github.getsmp.lorforandroid.ui.section.ItemFactory
import io.github.getsmp.lorforandroid.util.StringUtils
import org.jsoup.nodes.Element

class NewsItemFactory : ItemFactory {
    override fun prepareItems(body: Element, items: MutableList<Any>) {
        val articles = body.select("article")
        for (article in articles) {
            if (article.hasClass("mini-news")) {
                // Mini-news article
                items.add(MiniNewsItem(
                        url = article.select("a[href^=/news/]").first().attr("href").substring(1),
                        title = Html.fromHtml(article.select("a[href^=/news/]").first().ownText()).toString(),
                        commentsCount = Html.fromHtml(article.select("a").first().nextSibling().toString()).toString().replace("[()]".toRegex(), "")
                ))
            } else {
                // Standard article
                items.add(Item(
                        url = article.select("h2 > a[href^=/news/]").first().attr("href").substring(1),
                        title = Html.fromHtml(article.select("h2 > a[href^=/news/]").first().ownText()).toString(),
                        groupTitle = StringUtils.removeSectionName(article.select("div.group").first().text()),
                        tags = StringUtils.tagsFromElements(article.select("a.tag")),
                        date = article.select("time").first().ownText(),
                        author = article.select("a[itemprop^=creator], div.sign:contains(anonymous)").first()?.ownText()?.replace(" ()", ""),
                        comments = article.select("div.nav > a[href$=#comments]:eq(0)").first()?.ownText() ?: "Комментарии ограничены"
                ))
            }
        }
    }
}
